test/test_closed_streams.py

```diff
@@ -145,3 +104,199 @@ def test_closed_stream_not_present_in_streams_dict(self, frame_factory):
 
         # The streams dictionary should be empty.
         assert not c.streams
+
+
+class TestStreamsClosedByEndStream(object):
+    example_request_headers = [
+        (':authority', 'example.com'),
+        (':path', '/'),
+        (':scheme', 'https'),
+        (':method', 'GET'),
+    ]
+    example_response_headers = [
+        (':status', '200'),
+        ('server', 'fake-serv/0.1.0')
+    ]
+
+    @pytest.mark.parametrize(
+        "frame",
+        [
+            lambda self, ff: ff.build_data_frame(b'hello'),
+            lambda self, ff: ff.build_headers_frame(
+                self.example_request_headers, flags=['END_STREAM']),
+            lambda self, ff: ff.build_headers_frame(
+                self.example_request_headers),
+        ]
+    )
+    @pytest.mark.parametrize("clear_streams", [True, False])
+    def test_frames_after_recv_end_will_error(self,
+                                              frame_factory,
+                                              frame,
+                                              clear_streams):
+        """
+        A stream that is closed by receive END_STREAM will raise
+        ProtocolError when received unexpected frame.
+        """
+        c = h2.connection.H2Connection(client_side=False)
+        c.receive_data(frame_factory.preamble())
+        c.initiate_connection()
+
+        f = frame_factory.build_headers_frame(
+            self.example_request_headers, flags=['END_STREAM']
+        )
+        c.receive_data(f.serialize())
+        c.send_headers(
+            stream_id=1,
+            headers=self.example_response_headers,
+            end_stream=True
+        )
+
+        if clear_streams:
+            # Call open_outbound_streams to force connection clean
+            # closed stream
+            c.open_outbound_streams
```

Comment:
We should call `open_inbound_streams` here just to be safe: even though both share the same code path at the moment, they may diverge in future and we should try to force the correct behaviour.